src/iotjs_module_httpparser.cpp

```diff
@@ -0,0 +1,483 @@
+    return container->name(at, length);                                 \
+  }                                                                     \
+
+
+
+//http-parser callbacks
+HTTP_CB( OnMessageBegin );
+HTTP_DATA_CB( OnUrl );
+HTTP_DATA_CB( OnStatus );
+HTTP_DATA_CB( OnHeaderField );
+HTTP_DATA_CB( OnHeaderValue );
+HTTP_CB( OnHeadersComplete );
+HTTP_DATA_CB( OnBody );
+HTTP_CB( OnMessageComplete );
+
+
+const struct http_parser_settings settings = {
+  iotjs::OnMessageBegin,
+  iotjs::OnUrl,
+  iotjs::OnStatus,
+  iotjs::OnHeaderField,
+  iotjs::OnHeaderValue,
+  iotjs::OnHeadersComplete,
+  iotjs::OnBody,
+  iotjs::OnMessageComplete,
+};
+
+
+void HTTPParserWrap::Initialize(http_parser_type type) {
+  http_parser_init(&parser, type);
+  url.MakeEmpty();
+  status_msg.MakeEmpty();
+  n_fields = 0;
+  n_values = 0;
+  flushed = false;
+  had_exception = false;
+}
+
+
+JHANDLER_FUNCTION(Reinitialize, handler) {
+  IOTJS_ASSERT(handler.GetArgLength() == 1);
+  IOTJS_ASSERT(handler.GetThis()->IsObject());
+  IOTJS_ASSERT(handler.GetArg(0)->IsNumber());
+
+  JObject* jparser = handler.GetThis();
+
+  http_parser_type httpparser_type =
+    static_cast<http_parser_type>(handler.GetArg(0)->GetInt32());
+  IOTJS_ASSERT(httpparser_type == HTTP_REQUEST ||
+               httpparser_type == HTTP_RESPONSE);
+
+  HTTPParserWrap* parser =
+    reinterpret_cast<HTTPParserWrap*>(jparser->GetNative());
+  parser->Initialize(httpparser_type);
+
+  return true;
+}
+
+
+JHANDLER_FUNCTION(Finish, handler) {
+  IOTJS_ASSERT(handler.GetArgLength() == 0);
+  IOTJS_ASSERT(handler.GetThis()->IsObject());
+  JObject* jparser = handler.GetThis();
+  HTTPParserWrap* parser =
+    reinterpret_cast<HTTPParserWrap*>(jparser->GetNative());
+
+  int rv = http_parser_execute(&(parser->parser), &settings, NULL, 0);
+
+  if (rv != 0) {
+    enum http_errno err = HTTP_PARSER_ERRNO(&parser->parser);
+
+    JObject eobj(JObject::Error("Parse Error"));
+    JSETPROPERTY(eobj, "byteParsed", 0);
+    JSETPROPERTY(eobj, "code", http_errno_name(err));
+    handler.Return(eobj);
+  }
+
+  return true;
+}
+
+
+JHANDLER_FUNCTION(Execute, handler) {
+  IOTJS_ASSERT(handler.GetArgLength() == 1);
+  IOTJS_ASSERT(handler.GetThis()->IsObject());
+  IOTJS_ASSERT(handler.GetArg(0)->IsObject());
+
+  JObject* jparser = handler.GetThis();
+  HTTPParserWrap* parser =
+    reinterpret_cast<HTTPParserWrap*>(jparser->GetNative());
+
+
+  JObject* jbuffer = handler.GetArg(0);
+  BufferWrap* buffer = BufferWrap::FromJBuffer(*jbuffer);
+  char* buf_data = buffer->buffer();
+  int buf_len = buffer->length();
+
+
+  parser->cur_jbuf = jbuffer;
+  parser->cur_buf = buf_data;
+  parser->cur_buf_len = buf_len;
+  parser->had_exception = false;
+
+  int nparsed = http_parser_execute(&(parser->parser), &settings,
+                                    buf_data, buf_len);
+
+  parser->cur_jbuf = NULL;
+  parser->cur_buf = NULL;
+  parser->cur_buf_len = 0;
+
+
+  if (parser->had_exception) {
+    return false;
```

Comment:
I will delete the ```had_exception``` variable from ```HTTPParserWrap```. 
```had_exception``` is on when JS function call has issues. But it can be handled by ```MakeCallback``` throwing error routine.